File: src/main/java/top/cnuo/idlepool/gui/PoolGuiManager.java
package top.cnuo.idlepool.gui;

import top.cnuo.idlepool.integration.VisualBridge;
import top.cnuo.idlepool.pool.PoolDefinition;
import top.cnuo.idlepool.session.SessionManager;
import top.cnuo.idlepool.util.Messages;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryDragEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.Map;

public final class PoolGuiManager implements Listener {
    private static final int INFO_SLOT = 11;
    private static final int START_SLOT = 13;
    private static final int REWARDS_SLOT = 15;

    private final JavaPlugin plugin;
    private final VisualBridge visuals;
    private final SessionManager sessions;

    public PoolGuiManager(JavaPlugin plugin, VisualBridge visuals, SessionManager sessions) {
        this.plugin = plugin;
        this.visuals = visuals;
        this.sessions = sessions;
    }

    public void open(Player player, PoolDefinition pool) {
        Component title = Messages.parse(pool.displayName());
        PoolGuiHolder holder = new PoolGuiHolder(pool);
        Inventory inventory = Bukkit.createInventory(holder, 27, title);
        holder.attach(inventory);

        inventory.setItem(INFO_SLOT, item(
                pool.visuals().infoItem(), Material.BOOK,
                "gui.pool.info.name", "gui.pool.info.lore", Map.of()
        ));
        boolean active = sessions.isActive(player.getUniqueId());
        inventory.setItem(START_SLOT, item(
                pool.visuals().startItem(), active ? Material.RED_DYE : Material.LIME_DYE,
                active ? "gui.pool.stop.name" : "gui.pool.start.name",
                active ? "gui.pool.stop.lore" : "gui.pool.start.lore", Map.of()
        ));
        inventory.setItem(REWARDS_SLOT, item(
                pool.visuals().rewardsItem(), Material.CHEST,
                "gui.pool.rewards.name", "gui.pool.rewards.lore", Map.of("plan", pool.rewardPlan())
        ));
        player.openInventory(inventory);
    }

    @EventHandler
    public void onClick(InventoryClickEvent event) {
        if (!(event.getView().getTopInventory().getHolder(false) instanceof PoolGuiHolder holder)) {
            return;
        }
        event.setCancelled(true);
        if (!(event.getWhoClicked() instanceof Player player) || event.getRawSlot() != START_SLOT) {
            return;
        }
        if (sessions.isActive(player.getUniqueId())) {
            player.closeInventory();
            sessions.stop(player, false);
        } else {
            sessions.start(player, holder.pool());
        }
    }

    @EventHandler
    public void onDrag(InventoryDragEvent event) {
        if (event.getView().getTopInventory().getHolder(false) instanceof PoolGuiHolder) {
            event.setCancelled(true);
        }
    }

    private ItemStack item(
            String customId,
            Material fallback,
            String nameKey,
            String loreKey,
            Map<String, String> placeholders
    ) {
        ItemStack item = visuals.customItem(customId).orElseGet(() -> new ItemStack(fallback));
        ItemMeta meta = item.getItemMeta();
        meta.displayName(Messages.get(nameKey, placeholders));
        meta.lore(Messages.list(loreKey, placeholders));
        item.setItemMeta(meta);
        return item;
    }
}

File: src/main/kotlin/top/cnuo/idlepool/gui/PoolGuiManager.kt
package top.cnuo.idlepool.gui

import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryDragEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.plugin.java.JavaPlugin
import top.cnuo.idlepool.integration.VisualBridge
import top.cnuo.idlepool.pool.PoolDefinition
import top.cnuo.idlepool.session.SessionManager
import top.cnuo.idlepool.util.Messages

class PoolGuiManager(
    private val plugin: JavaPlugin,
    private val visuals: VisualBridge,
    private val sessions: SessionManager,
) : Listener {
    fun open(player: Player, pool: PoolDefinition) {
        val holder = PoolGuiHolder(pool)
        val inventory = Bukkit.createInventory(holder, 27, Messages.parse(pool.displayName))
        holder.attach(inventory)
        inventory.setItem(11, item(pool.visuals.infoItem, Material.BOOK, "gui.pool.info", emptyMap()))
        val active = sessions.isActive(player.uniqueId)
        inventory.setItem(13, item(pool.visuals.startItem, if (active) Material.RED_DYE else Material.LIME_DYE, if (active) "gui.pool.stop" else "gui.pool.start", emptyMap()))
        inventory.setItem(15, item(pool.visuals.rewardsItem, Material.CHEST, "gui.pool.rewards", mapOf("plan" to pool.rewardPlan)))
        player.openInventory(inventory)
    }

    @EventHandler fun onClick(event: InventoryClickEvent) {
        val holder = event.view.topInventory.getHolder(false) as? PoolGuiHolder ?: return
        event.isCancelled = true
        val player = event.whoClicked as? Player ?: return
        if (event.rawSlot != 13) return
        if (sessions.isActive(player.uniqueId)) { player.closeInventory(); sessions.stop(player, false) } else sessions.start(player, holder.pool)
    }
    @EventHandler fun onDrag(event: InventoryDragEvent) { if (event.view.topInventory.getHolder(false) is PoolGuiHolder) event.isCancelled = true }

    private fun item(customId: String, fallback: Material, key: String, placeholders: Map<String, String>): ItemStack =
        visuals.customItem(customId).orElseGet { ItemStack(fallback) }.apply {
            itemMeta = itemMeta.apply { displayName(Messages.get("$key.name", placeholders)); lore(Messages.list("$key.lore", placeholders)) }
        }
}
